File: javatoh/Step10/src/Tower.java
import java.util.*;

public class Tower {
	private String name;
	private ArrayList<Disk> disks;

	Tower(String name) {
		this.name = name;
		disks = new ArrayList<Disk>();
	}

	
	void addDisk(Disk d) throws Exception {
		
		if( this.disks.size() == 0) {
			this.disks.add(d);	// no disk exists, so any disk allowed.
			return;
		}
		
		// input_disk_size = get the input disk(d) size
		// topDis = get the top disk
		// topDiskSize = get top disk size
		// if input disk size > topDisSize
			// if input_disk_size > topDiskSize
				// we should throw exception
			// else
				// we can add
		int input_disk_size = d.getSize();
		
		int totalDisks = this.disks.size();
		int topDiskIndex = totalDisks-1;
		Disk topDisk = this.disks.get(topDiskIndex);
		int topDiskSize = topDisk.getSize();
		
		if( input_disk_size > topDiskSize) {
			Exception notAllowed;
			notAllowed = new Exception("Tower " + this.name + " does not allow " + input_disk_size + " to come on to " + topDiskSize);
			throw notAllowed;
		}
		
		// now allowed , so add it.
		this.disks.add(d);
	}
	
	void draw() {
		// we drew the pole
		drawPole();
		
		for(int i=this.disks.size()-1; i >=0 ; i--) {
			Disk d = this.disks.get(i);
			d.draw();
		}

		
		// draw base 
		drawBase();
		
		// title
		drawTitle();
	}

	Disk removeDisk() throws Exception{
		if( this.disks.size() == 0) {
			Exception noDisk;
			noDisk = new Exception("Tower: " + this.name + " does not have any disks");
			throw noDisk;
		}
		
		int totalDisks = this.disks.size();
		int topDiskIndex = totalDisks-1;
		Disk topDisk = this.disks.remove(topDiskIndex);
		return topDisk;
	}
	
	void drawPole() {
		for (int i = 0; i < 4; i++) {
			System.out.println("     |");
		}
	}

	private void drawBase() {
		System.out.println("________________");
	}

	private void drawTitle() {
		System.out.println("   " + this.name);
	}
}

File: pythontoh/Step9/tower.py
from disk import Disk


class Tower:

    # Constructor. which will be called during construction
    # which is Tower("some title")
    def __init__(self, title):  
        self.disk_bag = []
        self.title = title      # automatially adding a property to object

    def add_disk(self, disk):

        if len(self.disk_bag) == 0:
            self.disk_bag.append(disk)  # if no disks , we can add any disk
            return
        
        # get top disk size
        # get input disk size
        # if input disk size > top disk size
            # raise error
        total_disks = len(self.disk_bag)
        top_disk = self.disk_bag[total_disks-1]
        top_disk_size = top_disk.get_size()
        coming_disk_size = disk.get_size()

        if coming_disk_size > top_disk_size:
            raise ValueError(f"{self.title} disk:{coming_disk_size} cannot go on to {top_disk_size}")

        self.disk_bag.append(disk)

    def remove_disk(self):
        if len(self.disk_bag) == 0:
            raise ValueError(f'tower {self.title} does not have any disks to move.')

        # get the top disk
        total_disks = len(self.disk_bag)
        topDisk = self.disk_bag[total_disks-1]
        self.disk_bag.remove(topDisk)
        return topDisk

    def get_numberof_disks(self):
        return len(self.disk_bag)

    def drawPole(self):
        for x in range(1, 5, 1):
            print('                       |')

    def drawBase(self):
        print('           ---------------------')

    def drawTitle(self):
        print(f"                    {self.title}")

    def draw(self):
        # pole
        self.drawPole()

        # drawing disks
        for disk in self.disk_bag[::-1]:    # walking reverse
            disk.draw()

        # now base
        self.drawBase()

        # title now.
        self.drawTitle()
